sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -282,8 +282,12 @@ private[state] class HDFSBackedStateStoreProvider(
       // target file will break speculation, skipping the rename step is the only choice. It's still
       // semantically correct because Structured Streaming requires rerunning a batch should
       // generate the same output. (SPARK-19677)
+      // Also, a tmp file of delta file that generated by the first batch after restart
+      // streaming job is still reserved on HDFS. (SPARK-19779)
       // scalastyle:on
-      if (!fs.exists(finalDeltaFile) && !fs.rename(tempDeltaFile, finalDeltaFile)) {
+      if (fs.exists(finalDeltaFile)) {
+        fs.delete(tempDeltaFile, true)
+      } else if (!fs.rename(tempDeltaFile, finalDeltaFile)) {
```

Comment:
when restart streaming job , the`finalDeltaFile`  is same to a `finalDeltaFile` generated last batch of streaming job before restart. So here don't need rename to create an same file.